src/scheduler/queue.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 1994-2019 Altair Engineering, Inc.
+ * For more information, contact Altair at www.altair.com.
+ *
+ * This file is part of the PBS Professional ("PBS Pro") software.
+ *
+ * Open Source License Information:
+ *
+ * PBS Pro is free software. You can redistribute it and/or modify it under the
+ * terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option) any
+ * later version.
+ *
+ * PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "constant.h"
+#include "log.h"
+#include "queue.h"
+
+
+/**
+ * @brief	Constructor for the data structure 'queue'
+ *
+ * @param	void
+ *
+ * @return queue *
+ * @retval a newly allocated queue object
+ * @retval NULL for malloc error
+ */
+ds_queue *
+new_ds_queue(void)
```

Comment:
Can I recommend a change from ds_queue to thread_comm_queue or tc_queue. I'm not sure it needs 'ds' there in the name.

Reply:
So, this is a generic queue implementation, parts of the code other than threading can also make use of it. I added 'ds' in there because Bhroam was concerned that people might confuse this with PBS queues.

Reply:
In that case, how about renaming it to work_queue or pbs_queue?

Reply:
I think we should name it something completely separate.  pbs_queue is definitely wrong because pbs queues are what jobs are submitted into.  work_queue is the name of our default queue, so that might also be confusing.  BTW, Ravi, what does ds stand for?

Reply:
ds = data structure

I agree with bhroam that work_queue and pbs_queue can both be confusing.

you guys don't like ds_queue?

Reply:
@bhroam not really, I copied this data structure code from another project where I was using it as a normal queue, it's really just a generic queue.